README.md
# CRISPRi_designer
Tools to design gRNA/array pools targeting a list of genes. 

## Dependencies
Create a conda environment from the environment.yml file:
```
conda env create -f environment.yml
```

## get_sequences_from_gff.py
The first step in the pipeline is to obtain a fasta file with the sequences of the genes that are to be targeted. These sequences should contain the promoter. The file can be created from a gff annotation with the get_sequences_from_gff.py script.

### Usage
```
usage: get_sequences_from_gff.py [-h] [--outname OUTNAME]
                                 [--promoter PROMOTER] [--genetype GENETYPE]
                                 [--attribute ATTRIBUTE] [--name NAME]
                                 gff_file genome_file

Extracts gene sequences from gff and genome files and adds the promoter
sequence.

positional arguments:
  gff_file              Input annotation (gff) file with sequences of the
                        target genes.
  genome_file           Input genome file.

optional arguments:
  -h, --help            show this help message and exit
  --outname OUTNAME, -o OUTNAME
                        Base name of the output files. The .fasta extension
                        will be appended to this. (Default: target_genes.)
  --promoter PROMOTER, -p PROMOTER
                        Length before the gene start to be considered as the
                        promoter. Keep to 0 if you want to avoid targeting the
                        promoter or if you're not sure where the promoter is.
                        (Default: 0)
  --genetype GENETYPE, -t GENETYPE
                        Consider only genes of this type (Third column of the
                        gff file).
  --attribute ATTRIBUTE, -a ATTRIBUTE
                        Consider only genes that have this attribute (ninth
                        column of the gff file).E.g. if you want to filter for
                        the attribute "sRNA_type=Intergenic", write
                        sRNA_type=Intergenic.
  --name NAME, -n NAME  Take gene name from this attribute of the gff
                        attribute column (ninth column of the gff file).
```
### Input
* A gff annotation (*gff_file*) in the standard format (tab-separated, no header, columns: seqname,feature,start,end,score,strand,frame,attribute).
* A (multi)fasta file (*genome_file*) containing the genome file (chromosome(s) and plasmid(s), if present).

### Output
* A *outname*_coordinates.txt file containing the coordinates of each gene (promoter included).
* A *outname*_sequences.fasta file containing the genes' sequences (promoter included).

*outname* is defined by the --outname option.

## PAM_frequency.py
(Optional). This script scans the input sequences to find the frequency of selected PAM sequences among the input genes.

### Usage

```
usage: PAM_frequency.py [-h] [--input_file INPUT_FILE] [--PAMs PAMS]
                        [--spacer_length SPACER_LENGTH] [--binsize BINSIZE]

Calculates the frequency of each PAM in the input fasta file.

optional arguments:
  -h, --help            show this help message and exit
  --input_file INPUT_FILE, -i INPUT_FILE
                        Fasta file with sequences of the target genes +
                        promoters. (Default: target_genes_sequences.fasta).
  --PAMs PAMS, -pam PAMS
                        List of PAM sequences and their position relative to
                        the protospacer. Format: PAM1-position,PAM2-position
                        (can be more than two). (Default: TTV-5,NGG-3).
  --spacer_length SPACER_LENGTH, -sl SPACER_LENGTH
                        Length of each spacer. No more than 26 if the -a flag
                        is used. (Default: 20).
  --binsize BINSIZE, -b BINSIZE
                        Size of the bins in the final heatmap. (Default: 5).
```
### Input
* A fasta file (*input_file*) file containing the sequences of the target genes (promoter included).

### Output
* *PAM_frequency_counts.tsv*: tab-separated file with one PAM per row and number of genes that have 0...n occurrences of the PAM. Column names are the number of PAMs/gene.
* *PAM_frequency_heatmap.pdf*: Heatmap with the same results, binned in bins of size set by *--binsize*.

## design_CRISPRi_gRNAs.py
This script designs the gRNA/array library. PAM sequence (*--PAM*), strand preference inside the transcribed region (*PAM_preference*) and orientation respective to the protospacer (*--PAM_orientation*) can be freely set. gRNAs in the promoter can target both strands. Note that in the case of CDSs, the region immediately upstream the annotated start is not the promoter, but the 5-UTR and/or an upstream gene in the same operon. For these genes, unless the transcription start site is known, it's highly recommended to set *--promoter_length* to 0, to avoid gRNAs being designed for both strands on a region that is not the promoter.

For Cas12a, the script can design 1 array per gene if the *--arrays* flag is set. In this case, the oligonucleotides designed for the arrays are structured for cloning via CRATES (PMID: 31270316). The default *--left_overlap* and *--right_overlap* overlap sequences are for cloning into AWP-029.

This script supports multiprocessing (*--processors*), however >1 processors are only used when looking for offtargets and designing arrays (*--arrays*). The second case is the one where multiple processors really make a difference. With 40 processors, the script takes about 30 minutes for the ~150 genes of ~150 nt length.


### Usage

```
usage: design_CRISPRi_gRNAs.py [-h] [--input_file INPUT_FILE]
                               [--coordinates COORDINATES]
                               [--reference REFERENCE]
                               [--spacer_length SPACER_LENGTH]
                               [--promoter_length PROMOTER_LENGTH]
                               [--non_targeting [NON_TARGETING]] [--PAM PAM]
                               [--PAM_preference {template,nontemplate}]
                               [--PAM_orientation {5prime,3prime}]
                               [--spacers SPACERS] [--arrays]
                               [--folding FOLDING]
                               [--left_overlap LEFT_OVERLAP]
                               [--right_overlap RIGHT_OVERLAP]
                               [--processors PROCESSORS]

Designs spacers targeting the input genes.

optional arguments:
  -h, --help            show this help message and exit
  --input_file INPUT_FILE, -i INPUT_FILE
                        Fasta file with sequences of the target genes +
                        promoters. (Default: target_genes_sequences.fasta)
  --coordinates COORDINATES, -c COORDINATES
                        File with coordinates of the target genes. (Default:
                        target_genes_coordinates.txt)
  --reference REFERENCE, -r REFERENCE
                        Genome sequence file (Default: genome.fasta).
  --spacer_length SPACER_LENGTH, -sl SPACER_LENGTH
                        Length of each spacer. No more than 26 if the -a flag
                        is used. (Default: 20)
  --promoter_length PROMOTER_LENGTH, -pl PROMOTER_LENGTH
                        Length of the promoter region. Keep to 0 if you want
                        to avoid targeting the promoter or if you're not sure
                        where the promoter is. (Default: 0)
  --non_targeting [NON_TARGETING], -nt [NON_TARGETING]
                        The script designs nontargeting gRNAs if this flag is
                        present. If only -nt is set, the number of designed
                        nontargeting gRNAs is the largest number between 20
                        and total number of gRNAs divided by 100. If -nt n is
                        set, the script designs n nontargeting gRNAs.
  --PAM PAM, -pam PAM   PAM sequence. (Default: TTV)
  --PAM_preference {template,nontemplate}, -pp {template,nontemplate}
                        Which strand within the coding region should be
                        targeted? (Default: template)
  --PAM_orientation {5prime,3prime}, -po {5prime,3prime}
                        At which end of the protospacer is the PAM located?
                        (Default: 5prime)
  --spacers SPACERS, -s SPACERS
                        Number of spacers (including one array if using the -a
                        flag) to be designed per targeted gene. (Default: 4)
  --arrays, -a          The script attempts to design one array per target
                        gene if this flag is present. This flag can be used
                        only if --PAM is TTV.
  --folding FOLDING, -f FOLDING
                        Minimal accepted correct folding probability of the
                        arrays. Scale: 0-1. (Default: 0.2)
  --left_overlap LEFT_OVERLAP, -lo LEFT_OVERLAP
                        Left overhang for cloning grna spacers. (Default:
                        atctttgcagtaatttctactgttgtagat)
  --right_overlap RIGHT_OVERLAP, -ro RIGHT_OVERLAP
                        Right overhang for cloning grna spacers. (Default:
                        ccggcttatcggtcagtttcacctgattta)
  --processors PROCESSORS, -p PROCESSORS
                        Number of processors. A real impact of using >1
                        processors is seen only when also using the -a option.
                        (Default: 1)
```

### Input
The two files generated by _get_sequences_from_gff_ with sequences and coordinates of the genes to be targeted (plus the promoter). if different from the default, the file names can be indicated by *--input_file* and *--coordinates*. Please note that the gene names in *input_file* must not contain space characters.

### Output
The following files are generated by the script:
* *references.fasta*: Fasta file with all gRNAs (including nontargeting if the *-nt* option is used) and arrays (if the *-a* option is used).
* *gRNA_oligo_pool.txt*: text file with all gRNAs + homology arms (sequences only). Comprises the non targeting gRNAs too if the *-nt* option is used. Might be useful for ordering the oligonucleotides.
* *array_oligo_pool_n.txt*: only if the *-a* option is used. Text file(s) with the sequences of the oligos needed to construct the array via CRATES. One file is generated per each pool to be ordered.

## count_guides.py
This script makes a count table file with the counts of each gRNA/array in a fastq reads file. The input read file must be one (i.e. a unique file with merged mate reads if sequencing was paired-end).

### Usage

```
usage: count_guides.py [-h] [--reference REFERENCE] [--processors PROCESSORS] infile outfile

Count occurrences of gRNA/arrays in a read file.

positional arguments:
  infile                Input fastq file.
  outfile               Output counts file.

optional arguments:
  -h, --help            show this help message and exit
  --reference REFERENCE, -r REFERENCE
                        Input fastq file. (Default: references.fasta)
  --processors PROCESSORS, -p PROCESSORS
                        No. of processors. (Default: 1)

```
### Input
* A gzipped fastq file (*infile*) containing reads from a sequencing experiment.

### Output
* *outfile*: a tab-separated file with one colum with the name of the gRNA/array (from *reference*) and a second column with its count number.

## Examples

* Extract sequences of the targeted genes and promoters (of length 35 nt). The genome is genome.fasta, the gff file is annotation.gff. We want to filter the gff file for CDS entries only, while the name should be taken from the "gene_name" attribute:

	```
	get_sequences_from_gff.py -p 35 -t CDS -a gene_name annotation.gff genome.fasta 
	```

* Calculate frequencies of the following PAMs: TTV (5 prime of the protospacer), NGG (3 prime), NNNNGATT (3 prime) and NNGRR (3 prime). The target sequences are in targets.fasta and we want the final heatmap results to be binned in bins of size 3:

	```
	PAM_frequency.py -i targets.fasta -pam TTV-5,NGG-3,NNNNGATT-3,NNGRR-3 -b 3
	```

* Design a library of gRNAs, arrays and nontargeting gRNAs for Cas12a, with PAM TTV, template strand preference and 5prime orientation. The genome is genome.fasta and the target genes are found in target_genes_sequences.fasta and target_genes_sequences.fasta:

	```
	design_CRISPRi_gRNAs.py -i target_genes_sequences.fasta -c target_genes_sequences.fasta -r genome.fasta -a -nt -pam TTV -pp template -po 5prime
	```
 
* Design a library of gRNAs for Cas9, with PAM NGG, nontemplate strand preference and 3prime orientation. The genome is genome.fasta and the target genes are found in target_genes_sequences.fasta and target_genes_sequences.fasta:
	
	```
	design_CRISPRi_gRNAs.py -i target_genes_sequences.fasta -c target_genes_sequences.fasta -r genome.fasta -pam NGG -pp nontemplate -po 3prime
	```

* Design a library of targeting and nontargeting gRNAs for Cas9, with PAM NGG, nontemplate strand preference and 3prime orientation. We want 50 nontargeting gRNAs.
	
	```
	design_CRISPRi_gRNAs.py -nt 50 -pam NGG -pp nontemplate -po 3prime
	```

* Count the gRNAs and arrays of the file *references.fasta* in the fastq *reads.fastq* and save to *counts.tsv*. Use 20 processors.

	```
	python count_guides.py -p 20 -r references.fasta reads.fastq counts.tsv
	```

CHANGELOG.txt
v1.0.3 (2013-09-06)
- design_CRISPRi_gRNAs.py: fix offtarget search for 3prime PAMs
- PAM_frequency.py: add rotation of x labels in PAM plot
- update Readme
v1.0.2 (2013-08-02)
- design_CRISPRi_gRNAs.py: fix error when gene names contain an underscore.
- design_CRISPRi_gRNAs.py: get_sequences_from_gff.py: set default promoter length to 0.
- Update Readme.md
v1.0.1 (2013-07-27)
- design_CRISPRi_gRNAs.py: Fix spacer prioritizing by score, so that at parity of score, spacer closer to the gene's 5'end are picked.  
- design_CRISPRi_gRNAs.py: Fix some hardcoded spacer_length.
- Update README.md
- Add CHANGELOG.txt
v1.0.0 (2023-06-30)
- Published version, added PAM flexibility, strand-specificity option and many other things.
v0.1.0 (2013-06-29)
- Original version of the tools run to design the library used in the paper. Lots of hardcoded stuff, should not be used for any other purpose.

count_guides.py
import gzip
from Bio.SeqIO.QualityIO import FastqGeneralIterator
from Bio import SeqIO
from multiprocessing import Pool, Lock
from multiprocessing.managers import SharedMemoryManager
import argparse

parser = argparse.ArgumentParser(description='Count occurrences of gRNA/arrays in a read file.')

parser.add_argument('infile', type=str, help='Input fastq file.')
parser.add_argument('outfile', type=str, help='Output counts file.')
parser.add_argument('--reference', '-r', type=str, help='Input fastq file. (Default: references.fasta)',
					default = "references.fasta" )
parser.add_argument('--processors', '-p', type=int, help='No. of processors. (Default: 1)', default = 1)

args = parser.parse_args()
infile = args.infile
reference = args.reference
outfile = args.outfile
processors = args.processors

def count_refs(fastq_gen):
# counts each entry in references in the fastq file
	global counter_list
	[_, seq, __] = fastq_gen
	i = -1
	for [_,ref,ref_revc] in references:
		i+=1
		if ref in seq:
			lock.acquire()
			counter_list[i] += 1
			lock.release()
			break
		else:
			if ref_revc in seq:
				lock.acquire()
				counter_list[i] += 1
				lock.release()
				break

def revcomp(sequence):
# returns reverse complement of sequence
	complement = {'A':'T', 'C':'G', 'T':'A', 'G':'C', 'N':'N','Y':'R',
	'R':'Y', 'W':'W', 'S':'S', 'K':'M', 'M':'K', 'D':'H', 'H':'D',
	'V':'B', 'B':'V'}
	bases = list(sequence)
	bases.reverse()
	revcomp = ''
	for base in bases:
		if base.islower():
			base = base.upper()
			revcomp += complement[base].lower()
		else:
			revcomp += complement[base]
	return revcomp

#initialize list of references
references = []
for record in SeqIO.parse(reference,"fasta"):
	references.append([record.id,str(record.seq.upper()),revcomp(str(record.seq.upper()))])

# count gRNAs:
lock = Lock()
with SharedMemoryManager() as smm:
	counter_list = smm.ShareableList([0 for x in references])
	in_handle = gzip.open(infile,"rt")
	pool=Pool(processors)
	fastq_gen = FastqGeneralIterator(in_handle)
	pool.map(count_refs, fastq_gen)
	pool.close()
	pool.join()
	in_handle.close()
	smm.shutdown()

#write results to outfile:
with open(outfile, mode="w") as outf:
	for ref,count in zip(references,list(counter_list)):
		print(ref[0],count)
		outf.write("\t".join([ref[0],str(count)+"\n"]))

get_sequences_from_gff.py
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from collections import defaultdict
import argparse
import sys

parser = argparse.ArgumentParser(description='Extracts gene sequences from gff and genome files and adds the promoter sequence.')

parser.add_argument('gff_file', type=str,
					help='Input annotation (gff) file with sequences of the target genes.')
					
parser.add_argument('genome_file', type=str,
					help='Input genome file.')

parser.add_argument('--outname', '-o', type=str, default="target_genes",
					help='Base name of the output files.'
					'(Default: target_genes.)')

parser.add_argument('--promoter', '-p', default=0, type=int,
					help='Length before the gene start to be considered as the promoter. Keep to 0 if you want to avoid targeting '
					'the promoter or if you\'re not sure where the promoter is. (Default: 0)')

parser.add_argument('--genetype', '-g', type=str,
					help='Consider only genes of this type (Third column of the gff file).')			

parser.add_argument('--attribute', '-a', type=str,
					help='Consider only genes that have this string in the attributes column (ninth column of the gff file).'
					'E.g. if you want to filter for the attribute "sRNA_type=Intergenic", write '
					'sRNA_type=Intergenic.')

parser.add_argument('--name', '-n', type=str, default="locus_tag",
					help='Take gene name from this attribute of the gff attribute column (ninth column of the gff file).')	
					
args = parser.parse_args()
gff_file = args.gff_file
genome_file = args.genome_file
outname = args.outname
promoter_length = args.promoter
genetype = args.genetype
attribute = args.attribute
gene_name = args.name

def revcomp(sequence):
#reverse complements sequence
	basecomplement = {'A':'T', 'C':'G', 'T':'A', 'G':'C'} 
	letters = list(sequence) 
	letters.reverse() 
	dna = ''
	for base in letters:
		dna += basecomplement[base] 
	return dna

##### extend sRNA coordinates to add promoter_length before the start:
input_file = open(gff_file, mode="r")
genes = {}
for line in input_file:
	linelist = line.split("\t")
	if line.startswith("#") or not line.rstrip():
		continue
	if genetype is not None:
		if linelist[2] != genetype:
			continue
	if attribute is not None:
		if attribute not in linelist[8]:
			continue
	seqname = linelist[0]
	attributes = {}
	for attr in linelist[8].split(";"):
		attr = attr.split("=")
		attributes[attr[0]] = attr[1]
	try:
		gene = attributes[gene_name].rstrip()
	except KeyError:
		print("ERROR: this line of the gff file: \n\n{}\ndoesn't have the "
		"desired attribute '{}' which is set by the -n flag.".format(line,gene_name))
		sys.exit()
	strand = linelist[6]
	if strand == "+":
		start = str(int(linelist[3]) - promoter_length)
		end = str(int(linelist[4]))
	elif strand == "-":
		end = str(int(linelist[4]) + promoter_length)
		start = str(int(linelist[3]))
	genes[gene] = [seqname,start,end,strand]

with open("{}_coordinates.txt".format(outname), mode="w") as outf:
	for gene,[scaffold,start,end,strand] in genes.items():
		line = "\t".join([scaffold,gene,start,end,strand])
		outf.write(line)
		outf.write("\n")

##### make a fasta file with gene name and sequence:
input_file = open(genome_file)
#make a dictionary out of the genome:
assembly_seqs = SeqIO.to_dict(SeqIO.parse(input_file, "fasta"))
#make a list of SeqRecord items of each gene:
genes_sr = []
for gene,coord in genes.items():
	[seqname,start,end,strand] = coord
	gene_sequence = assembly_seqs[seqname][int(start)-1:int(end)].seq.upper()
	if strand == "-":
		gene_sequence = revcomp(gene_sequence)
	gene_sr = SeqRecord(gene_sequence, id=gene, description ="")
	genes_sr.append(gene_sr)
	genes[gene].append(gene_sequence)

#make a dictionary of sequence:geneID for each gene.
#this is to capture genes that have the same sequence
#and store their IDs as a list in the dictionary value
seq_dict = defaultdict(list)
for gene in genes_sr:
	seq_dict[str(gene.seq)].append(gene.id)

#write the output fasta file:
out_fasta = open("{}_sequences.fasta".format(outname), 'w')
for seq, IDs in seq_dict.items():
	#Join the IDs of genes that have the same sequence as gene1|gene2:
	ID = '|'.join(IDs)
	out_fasta.write(">{}\n".format(ID))
	out_fasta.write(seq + "\n")

PAM_frequency.py
from Bio import SeqIO
from Bio import Data
from itertools import product
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
matplotlib.use('agg')
from collections import OrderedDict
import argparse

parser = argparse.ArgumentParser(description='Calculates the frequency of each PAM in the input fasta file.')

parser.add_argument('--input_file', '-i', type=str, default="target_genes_sequences.fasta",
					help='Fasta file with sequences of the target genes + promoters. (Default: target_genes_sequences.fasta).')

parser.add_argument('--PAMs', '-pam', default="TTV-5,NGG-3", type=str, help='List of PAM sequences and their position '
					'relative to the protospacer. Format: PAM1-position,PAM2-position (can be more than two). (Default: TTV-5,NGG-3).')

parser.add_argument('--spacer_length', '-sl', default="20", type=int,
					help='Length of each spacer. No more than 26 if the -a flag is used. (Default: 20).')

parser.add_argument('--binsize', '-b', default="5", type=int,
					help='Size of the bins in the final heatmap. (Default: 5).')

args = parser.parse_args()
input_file = args.input_file
PAM_args = args.PAMs
spacer_length = args.spacer_length
binsize = args.binsize

ambiguous_alph = {"N":["A","C","G","T"], "V":["A","C","G"], "H":["A","C","T"], "D":["A","G","T"],
	"B":["C","G","T"], "M":["A","C"], "K":["G","T"], "W":["A","T"], "S":["C","G"], "Y":["C","T"],
	"R":["A","G"], "A": ["A"], "C": ["C"], "G": ["G"], "T": ["T"]}

def remove_ambiguous(sequence):
# returns a list of all possible sequences
# given the ambiguous sequence input (e.g. with N,V, etc. bases)
	return ["".join(i) for i in product(*[ ambiguous_alph[j] for j in sequence ]) ]

def revcomp(sequence):
#reverse complements sequence
	complement = {'A':'T', 'C':'G', 'T':'A', 'G':'C', 'N':'N','Y':'R',
	'R':'Y', 'W':'W', 'S':'S', 'K':'M', 'M':'K', 'D':'H', 'H':'D',
	'V':'B', 'B':'V'} 
	bases = list(sequence) 
	bases.reverse() 
	revcomp = ''
	for base in bases:
		revcomp += complement[base] 
	return revcomp

def find_all(string, substr):
# makes a list with the indexes of all occurrences of substr in string
	listindex=[]
	start = 0
	i = string.find(substr, start)
	while i >= 0:
		listindex.append(i)
		i = string.find(substr, i + 1)
	return listindex	

def find_guides(seq):
# finds gRNAs in seq
	guides_bottom = []
	guides_top = []
	for PAM_sequence in PAM_list:
		guides_top.extend(find_all(seq, revcomp(PAM_sequence)))
		guides_bottom.extend(find_all(seq, PAM_sequence))
	return guides_top, guides_bottom

def binned_sum(data):
# Sums the values in data in bins of binsize size. Returns an OrderedDict
# where keys are the bin number (starting from 1) and values the sum of the
# values in that bin. The first item (key=0) equals to data[0].
	binned_data = OrderedDict()
	binned_data[0] = data[0]
	j = 0
	for i in bins:
		binned_data[j+1] = 0
		for k in range(bins[j]-binsize+1,bins[j]+1):
			if k in data:
				binned_data[j+1] += data[k]
		j += 1
		if j == len(bins):
			return binned_data
	return binned_data

#create a list of the PAMs in the style [["TTV",5],["NGG",3]]

PAM_args = PAM_args.split(",")
PAMs = []
for PAM in PAM_args:
	seq,pos = PAM.split("-")
	pos = int(pos)
	if pos not in (3, 5):
		print("ERROR: The PAM position you gave to the --PAMs flag is invalid. It has to be either 3 or 5.")
		exit()
	PAMs.append([seq,pos])

#make a PAMs_dic dictionary with each PAM as key and as value a dictionary with 
#key: number of gRNAs/gene value:number of gRNAs that have that number
PAMs_dic = {}
max_no = 0
for PAM in PAMs:
	spacer_count_dict = OrderedDict()
	sRNA_list = SeqIO.parse(input_file, 'fasta')
	PAM_list = remove_ambiguous(PAM[0])
	PAM_length = len(PAM_list[0])
	for sRNA in sRNA_list:
		sRNA_sequence = str(sRNA.seq)
		sRNA_length = len(sRNA_sequence)
		guides_top, guides_bottom = find_guides(sRNA_sequence)
		gRNA_number = 0
		if PAM[1] == 3: #eg NGG
			for pos in guides_top:
				if pos <= sRNA_length - spacer_length - PAM_length:
					gRNA_number += 1
			for pos in guides_bottom:
				if pos >= spacer_length:
					gRNA_number += 1
		else: #eg TTV
			for pos in guides_bottom:
				if pos <= sRNA_length - spacer_length - PAM_length:
					gRNA_number += 1
			for pos in guides_top:
				if pos >= spacer_length:
					gRNA_number += 1
		if gRNA_number in spacer_count_dict:
			spacer_count_dict[gRNA_number] += 1
		else:
			spacer_count_dict[gRNA_number] = 1
	PAMs_dic[PAM[0]] = spacer_count_dict
	max_no = max(max_no,max(spacer_count_dict))

#add an entry of 0 for each value that is missing
for PAM, counts in PAMs_dic.copy().items():
	for i in range(0,max_no+1):
		if i not in counts:
			PAMs_dic[PAM][i] = 0

#Create the the results table file
outfile = open("PAM_frequency_counts.tsv",mode="w+")
outfile.write("\t".join(["PAM"] + [str(x) for x in range(max_no+1)]))
outfile.write("\n")

#Bin the dictionaries in PAMs_dic into bins of binsize
#at the same time, write to file
num_bins = max_no // binsize + 1
bins = [i * binsize for i in range(1,num_bins+1)]
max_no_binned = 0
for PAM, counts in PAMs_dic.copy().items():
	ordered_counts = OrderedDict(sorted(counts.items(),reverse=False))
	outfile.write("\t".join([PAM] + [str(x) for x in ordered_counts.values()]))
	outfile.write("\n")
	PAMs_dic[PAM]=binned_sum(ordered_counts)
	max_no_binned = max(max_no_binned,max(PAMs_dic[PAM].values()))

#prepare lists for the heatmap
x_values = [] #x values in the final heatmap
y_values = [] #y values in the final heatmap
y_labels = [] #y_labels in the final heatmap
x_labels = [str(0)] #x_labels in the final heatmap

for PAM in PAMs:
	pam = PAM[0]
	x_values.append(list(PAMs_dic[pam].keys()))
	y_values.append(PAMs_dic[pam].values())
	y_labels.append(pam)
y_values=[list(y) for y in y_values]

for i in bins:
	x_labels.extend(["{}-{}".format(i-binsize+1,i)])

# Create the heatmap
plt.rcParams.update({"font.size": 13, "xtick.labelsize": 13, "ytick.labelsize": 13, "axes.labelsize": 13})
plt.figure(figsize=(len(x_labels), len(y_labels)))
y = np.array(y_values)
heatmap = plt.pcolormesh(y, cmap=plt.cm.Blues, vmax=max_no_binned)

# Add annotations
for i in range(len(y_labels)):
    for j in range(len(x_labels)):
        plt.text(j + 0.5, i + 0.5, str(y[i, j]), ha='center', va='center')

# Customize, ticks, labels and titles
plt.yticks(np.arange(len(y_labels)) + 0.5, y_labels)
plt.xticks(np.arange(len(x_labels)) + 0.5, x_labels, rotation=30, ha='right',rotation_mode='anchor')
plt.colorbar(label='Number of gRNAs')
plt.xlabel("Number of PAMs/gene")
plt.gca().invert_yaxis()

# Adjust the plot limits
plt.xlim(0, len(x_labels))
plt.ylim(0, len(y_labels))

# Add boundary lines
plt.axhline(y=0, color='0.2', linewidth=2)
plt.axhline(y=len(y_labels), color='0.2', linewidth=2)
plt.axvline(x=0, color='0.2', linewidth=2)
plt.axvline(x=len(x_labels), color='0.2', linewidth=2)

plt.savefig('PAM_frequency_heatmap.pdf', bbox_inches='tight')